### ISSUE TITLE
Narrowing IO[str] | None with sys.stdout is still considered possibly None

### ISSUE THREAD

**Issue by nedbat** [microsoft/pyright#11768]
**Describe the bug**
A clear and concise description of the behavior you are seeing and the expected behavior along with steps to reproduce it.

**Code or Screenshots**
If possible, provide a minimal, self-contained code sample (surrounded by triple back ticks) to demonstrate the issue. The code should define or import all referenced symbols.

```python
import sys
from typing import IO, reveal_type

def report(outfile: IO[str] | None = None) -> None:
    reveal_type(sys.stdout)
    outfile = outfile or sys.stdout
    reveal_type(outfile)
    outfile.write("Hello")
```

This produces (with pyright 1.1.414):
```
/Users/nbatchelder/n/coveragepy/prbug.py
  /Users/nbatchelder/n/coveragepy/prbug.py:5:17 - information: Type of "sys.stdout" is "TextIO | Any"
  /Users/nbatchelder/n/coveragepy/prbug.py:7:17 - information: Type of "outfile" is "IO[str] | TextIO | None"
  /Users/nbatchelder/n/coveragepy/prbug.py:8:13 - error: "write" is not a known attribute of "None" (reportOptionalMemberAccess)
1 error, 0 warnings, 2 informations
```

outfile cannot be None on line 8.

**Comment by rchiodo** [comment]
This is expected behavior rather than a bug in `or` narrowing. The important detail is the `Any` in the revealed type of `sys.stdout`.

Pyright supports narrowing with `or`. In `outfile or sys.stdout`, it excludes `None` from the left operand's contribution to the result, because a falsy left operand causes the right operand to be returned. However, that does not establish that the right operand is non-`None`.

Typeshed declares `sys.stdout` as `TextIO | MaybeNone`, where `MaybeNone` is an alias for `Any`. This is the documented ["Any trick"](https://typing.python.org/en/latest/guides/writing_stubs.html#the-any-trick): it represents a possible `None` without requiring a check at every direct use of the API.

When the expression is assigned back to `outfile`, whose declared type is `IO[str] | None`, Pyright's assignment-narrowing rules retain declared union members compatible with the assigned `Any`. That includes `None`, which explains the resulting `IO[str] | TextIO | None`. The `None` is therefore not evidence that narrowing of the left operand failed; it remains possible through `sys.stdout`.

In fact, `outfile` can be `None` at runtime here. [Python documents](https://docs.python.org/3/library/sys.html#sys.__stdout__) that `sys.stdout` can itself be `None`, notably for Windows GUI applications without a console and applications started with `pythonw`. If both operands are `None`, the result is `None`.

An explicit check handles this case and narrows the type:

```python
outfile = outfile or sys.stdout
if outfile is None:
    raise RuntimeError("No output stream is available")
outfile.write("Hello")
```

This is not a typing-spec restriction on `or`: [Pyright documents support for it](https://github.com/microsoft/pyright/blob/main/docs/type-concepts-advanced.md#type-guards), while the [typing spec](https://typing.python.org/en/latest/spec/narrowing.html) leaves general narrowing largely unspecified. The exact inferred union reflects Pyright's assignment-narrowing rules.

**Comment by nedbat** [comment]
The entire point of typeshed's MaybeNone type is to keep people from having to check for None. Please consider a way to honor their design.

**Comment by rchiodo** [comment]
I believe the root cause is `or`ng Any with None. It comes out as being possibly None. It's not the MaybeNone that's the problem, it's the `or` with None.

This code reproduces the same problem:

```python
def report(outfile: IO[str] | None = None) -> None:
    x: Any = 1
    reveal_type(x)
    outfile = outfile or x
    reveal_type(outfile)
    outfile.write("Hello")
```

**Comment by zsimic** [comment]
The handling of the assignment is at play here, here's a probe that will evidence it.
mypy, ty, pyrefly are OK with this, but pyright isn't, and only for line 9, when using the re-assigned `outfile`. Using the fresh new name `outfile2` is not reported as an issue by pyright

```
import sys
from typing import IO, reveal_type

def report(outfile: IO[str] | None = None) -> None:
    outfile2 = outfile or sys.stdout
    outfile = outfile or sys.stdout
    reveal_type(outfile)
    reveal_type(outfile2)
    outfile.write("Hello")
    outfile2.write("Hello")
```

**Comment by rchiodo** [comment]
That inconsistency is weird.
